src/coreclr/jit/lower.cpp

```diff
@@ -8015,32 +8077,84 @@ void Lowering::LowerStoreIndirCoalescing(GenTreeStoreInd* ind)
 
             case TYP_SHORT:
             case TYP_USHORT:
-                newType = TYP_INT; // TYP_UINT is not legal in IR
+                newType = TYP_INT;
                 break;
 
 #ifdef TARGET_64BIT
             case TYP_INT:
                 newType = TYP_LONG;
                 break;
+
+#if defined(FEATURE_HW_INTRINSICS)
+            case TYP_LONG:
+            case TYP_REF:
+                if (comp->IsBaselineSimdIsaSupported())
+                {
+                    // TLDR: we should be here only if one of the conditions is true:
+                    // 1) Both GT_INDs have GTF_IND_ALLOW_NON_ATOMIC flag
+                    // 2) ARM64: Data is at least 8-byte aligned
+                    // 3) AMD64: Data is at least 16-byte aligned on AMD/Intel with AVX+
+                    //
+                    newType = TYP_SIMD16;
+                    if ((oldType == TYP_REF) &&
+                        (!currData.value->IsIntegralConst(0) || !prevData.value->IsIntegralConst(0)))
+                    {
+                        // For TYP_REF we only support null values. In theory, we can also support frozen handles, e.g.:
+                        //
+                        //   arr[1] = "hello";
+                        //   arr[0] = "world";
+                        //
+                        // but we don't want to load managed references into SIMD registers (we can only do so
+                        // when we can issue a nongc region for a block)
+                        return;
+                    }
+                    break;
+                }
+                return;
+
+#if defined(TARGET_AMD64)
+            case TYP_SIMD16:
+                // Check for AVX to present and PreferredVectorByteLength
+                if (comp->compOpportunisticallyDependsOn(InstructionSet_AVX) &&
+                    (comp->getPreferredVectorByteLength() >= 32))
```

Comment:
Are both checks necessary?

Reply:
The first one check whether ISA is available to use. The 2nd checks whether we're allowed to use vectors of the given width. `comp->compOpportunisticallyDependsOn(InstructionSet_AVX)` doesn't take `PreferredVectorByteLength` into account because there 128-bit/scalar instructions they can provide which still can be used.

Reply:
I see, I thought `getPreferredVectorByteLength() == 32` implied support for the ISA

Reply:
UPD: it does, I was wrong